frame/uniques/src/lib.rs

```diff
@@ -0,0 +1,1182 @@
+				.ok_or(Error::<T, I>::Unknown)?;
+			let class_details = Class::<T, I>::get(&class).ok_or(Error::<T, I>::Unknown)?;
+			ensure!(class_details.admin == origin, Error::<T, I>::NoPermission);
+
+			details.is_frozen = false;
+			Asset::<T, I>::insert(&class, &instance, &details);
+
+			Self::deposit_event(Event::<T, I>::Frozen(class, instance));
+			Ok(())
+		}
+
+		/// Disallow further unprivileged transfers for a whole asset class.
+		///
+		/// Origin must be Signed and the sender should be the Freezer of the asset `class`.
+		///
+		/// - `class`: The asset class to be frozen.
+		///
+		/// Emits `ClassFrozen`.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::freeze_class())]
+		pub(super) fn freeze_class(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			Class::<T, I>::try_mutate(class, |maybe_details| {
+				let details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+				ensure!(&origin == &details.freezer, Error::<T, I>::NoPermission);
+
+				details.is_frozen = true;
+
+				Self::deposit_event(Event::<T, I>::ClassFrozen(class));
+				Ok(())
+			})
+		}
+
+		/// Re-allow unprivileged transfers for a whole asset class.
+		///
+		/// Origin must be Signed and the sender should be the Admin of the asset `class`.
+		///
+		/// - `class`: The class to be thawed.
+		///
+		/// Emits `ClassThawed`.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::thaw_class())]
+		pub(super) fn thaw_class(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			Class::<T, I>::try_mutate(class, |maybe_details| {
+				let details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+				ensure!(&origin == &details.admin, Error::<T, I>::NoPermission);
+
+				details.is_frozen = false;
+
+				Self::deposit_event(Event::<T, I>::ClassThawed(class));
+				Ok(())
+			})
+		}
+
+		/// Change the Owner of an asset class.
+		///
+		/// Origin must be Signed and the sender should be the Owner of the asset `class`.
+		///
+		/// - `class`: The asset class whose owner should be changed.
+		/// - `owner`: The new Owner of this asset class.
+		///
+		/// Emits `OwnerChanged`.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::transfer_ownership())]
+		pub(super) fn transfer_ownership(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId,
+			owner: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			let owner = T::Lookup::lookup(owner)?;
+
+			Class::<T, I>::try_mutate(class, |maybe_details| {
+				let details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+				ensure!(&origin == &details.owner, Error::<T, I>::NoPermission);
+				if details.owner == owner {
+					return Ok(());
+				}
+
+				let deposit = details.deposit + details.metadata_deposit;
```

Comment:
```suggestion
				let deposit = details.deposit.saturating_add(details.metadata_deposit);
```

Reply:
Unneeded - we know for certain that it doesn't overflow since `reserve` would have failed otherwise.